File: perl/str_reverse.pl
#!/usr/bin/perl

use strict;
use warnings;

use Data::Dumper;
use Encode;

my $str = 'abcde';
print Dumper(scalar reverse($str));

$str = '日本語表示';
print Dumper(str_reverse($str));

sub str_reverse {
    my ($str) = shift;
    utf8::decode($str); # utf8フラグをつける
    my @str = map {utf8::encode($_); $_} split //, $str;
    my $b   = @str;
    for (my $a = 0; $a <= --$b; $a++) {
      ($str[$a], $str[$b]) = ($str[$b], $str[$a]);
    }
    join('', @str);
}
__END__

File: python/str_reverse.py
#!/usr/local/bin/python
#-*- coding:utf8 -*-

str = u'あいうえお'

def str_reverse(str):
    li = list(str)
    l = len(str)
    for i in range(len(li) / 2):
        l = l - 1
        li[i], li[l] = li[l], li[i]
    return ''.join(li)

print str_reverse(str)

